### PR TITLE
Eos 24091: Changes in vkvv implementation for multihreaded UTs

### PR DESCRIPTION
**Pull request by ydb242**

### CODE REVIEW

File: btree/btree.c [hunk 1]
```diff
@@ -4820,15 +4827,34 @@ static void vkvv_make(struct slot *slot)
...
+	int               diff;
 
 	M0_PRE(slot->s_idx < h->vkvv_used && h->vkvv_used > 0);
 
+	if (vsize_diff > 0 &&
+	    (end_val_addr - end_dir_addr) < vsize_diff) {
+		diff = vsize_diff - (end_val_addr - end_dir_addr);
+
+		if (start_dir_addr - end_key_addr < diff ||
+		    (end_val_addr - end_dir_addr) < 0)
```
Comment (shashank-parulekar) [inline btree/btree.c:4850]: Why do you need this 2nd condition (end_val_addr - end_dir_addr) < 0  ??

**Review by shashank-parulekar — CHANGES_REQUESTED** [review]

File: btree/btree.c (hunk 1 above)
Comment (ydb242) [reply btree/btree.c:4850]: This particular condition was used for testing. I will remove it in the upcoming update.

File: btree/btree.c [hunk 2]
```diff
@@ -4233,9 +4233,14 @@ static int  vkvv_space(const struct nd *node)
 	if (h->vkvv_level == 0) {
 		rec                = vkvv_get_dir_addr(node);
```
[the remark this answers is not in the document]
Comment (ydb242) [reply btree/btree.c]: Updated 'rec' with 'ref' to denote directory. The rationale behind the usage of this variable name is to indicate the function of directory i.e. a reference list for keys and values.

File: btree/btree.c [hunk 3]
```diff
@@ -4233,9 +4233,14 @@ static int  vkvv_space(const struct nd *node)
...
 		dir_size           = (sizeof(struct dir_rec)) *
-				     (h->vkvv_used + 1);
-		size_of_all_keys   = rec[h->vkvv_used].key_offset;
-		size_of_all_values = rec[h->vkvv_used].val_offset;
+				     (h->vkvv_used + 2);
```
[the remark this answers is not in the document]
Comment (ydb242) [reply btree/btree.c]: Updated

File: btree/btree.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (shashank-parulekar) [reply btree/btree.c]: Use a name such as dir_entry which makes it more obvious instead of ref.

[the remark this answers is not in the document]
Comment (ydb242) [reply btree/btree.c]: Made the changes

File: btree/btree.c [hunk 4]
```diff
@@ -4338,8 +4343,7 @@ static void vkvv_fid(const struct nd *node, struct m0_fid *fid)
 
 static uint32_t vkvv_lnode_rec_key_size(const struct nd *node, int idx)
 {
-	struct vkvv_head *h   = vkvv_data(node);
-	struct dir_rec   *rec = (void *)h + sizeof(*h) + h->vkvv_dir_offset;
+	struct dir_rec *rec = vkvv_get_dir_addr(node);
 
 	return rec[idx+1].key_offset - rec[idx].key_offset;
```
[the remark this answers is not in the document]
Comment (ydb242) [reply btree/btree.c]: Done

**Review by RadhaGulhane13 — APPROVED** [review]

**Review by NikhilBirgade — APPROVED** [review]

**Review by shashank-parulekar — APPROVED** [review]

**Pull request merged**